pkg/controller/instance/instance_controller.go

```diff
@@ -414,36 +281,37 @@ func getOperatorVersion(ctx context.Context, c client.Client, r record.EventReco
 			instance.Spec.OperatorVersion.Name,
 			instance.Name,
 			err)
-		if r != nil {
-			r.Event(instance, "Warning", "InvalidOperatorVersion", fmt.Sprintf("Error getting operatorversion \"%v\": %v", ov.Name, err))
-		}
+		r.Recorder.Event(instance, "Warning", "InvalidOperatorVersion", fmt.Sprintf("Error getting operatorversion \"%v\": %v", instance.Spec.OperatorVersion.Name, err))
 		return nil, err
 	}
 	return ov, nil
 }
 
-// getInstance retrieves the instance by namespace name
-func getInstance(ctx context.Context, r *ReconcileInstance, request reconcile.Request) (instance *kudov1alpha1.Instance, err error) {
-	instance = &kudov1alpha1.Instance{}
-	err = r.Get(ctx, request.NamespacedName, instance)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
-			return nil, nil
+func getParameters(instance *kudov1alpha1.Instance, operatorVersion *kudov1alpha1.OperatorVersion) (map[string]string, error) {
+	params := make(map[string]string)
+
+	for k, v := range instance.Spec.Parameters {
+		params[k] = v
+	}
+
+	missingRequiredParameters := make([]string, 0)
+	// Merge defaults with customizations
+	for _, param := range operatorVersion.Spec.Parameters {
+		_, ok := params[param.Name]
+		if !ok && param.Required && param.Default == nil {
+			// instance does not define this parameter and there is no default while the parameter is required -> error
+			missingRequiredParameters = append(missingRequiredParameters, param.Name)
+
+		} else if !ok {
+			params[param.Name] = kudo.StringValue(param.Default)
 		}
-		// Error reading the object - requeue the request.
-		return nil, err
 	}
-	log.Printf("InstanceController: Received Reconcile request for instance \"%+v\"", request.Name)
-	return instance, nil
-}
 
-// isNewInstance detects if the instance does NOT have plan
-func isNewInstance(instance *kudov1alpha1.Instance) bool {
-	// if ActivePlan.Name is empty the instance is being created.  The instance will forever
-	// after have an active plan (which may be complete)
-	return instance.Status.ActivePlan.Name == ""
+	if len(missingRequiredParameters) != 0 {
+		return nil, &executionError{err: fmt.Errorf("parameters are missing when evaluating template: %s", strings.Join(missingRequiredParameters, ",")), fatal: true, eventName: kudo.String("Missing parameter")}
```

Comment:
we should consider creating a Kubernetes event here as well.

Reply:
we actually do! :) the execution error contains name of the event and that is then emitted "automatically" in `handleError` so the error type already holds the event information